### DISCUSSION TITLE
Complete Migration from Plexus to Eclipse Sisu

### DISCUSSION THREAD

**Discussion by sparsick** [support-and-care/maven-support-and-care#28] [Ideas for Funding]
A long time ago, Maven switched from Plexus to Eclipse Sisu.

My understanding is that it is a core functionality because it is the dependency injection framework (I hope, I'm correct). 

Ideas:
- Let's ensure that plexus is not used in core and plugins, that are managed by Apache (if not already done)
- Support for 3rd party plugins (tutorial, rewrite recipe etc.)
- Clean up documentation where Plexus is mentioned.
- Migration of elder plugin outside Apache (for example Mojohaus)

**Comment by rmannibucau** [comment]
+1, the plugins (and extensions) must use v4 api now which totally hides the maven "core" internal IoC. Work started but is not yet complete.

**Comment by khmarbaise** [comment]
Technically Maven can use https://maven.apache.org/maven-jsr330.html (JSR 330) which means using JSR330 compliant annotations (https://github.com/eclipse/sisu.plexus/wiki/Plexus-to-JSR330) should be documented better and examples should be created/updated accordingly.

**Reply by rmannibucau** [reply]
Hmm, technically JSR330 does not define any sufficient behavior and is already legacy so hope we don't promote it at all for external use since we would ultimately 1. defeat one of the best mvn4 purpose and 2. just expose Guice and not JSR330 (the API is just 10% of the contract there).

**Reply by khmarbaise** [reply]
An plain Guice would be also an option... Yes there are a number of limitations that's true...

**Reply by kwin** [reply]
I don't think plain Guice works well, as it requires a lot of glue code. That was the initial driver for having Sisu in the first place. In the best case the DI framework works with annotations only/mostly.

**Reply by rmannibucau** [reply]
Not sure it is really a topic now maven has its own official, stable and public DI API, only real question is when guice will be dropped. Sounds unlikely short term but even more unlikely we couple more with guice too. Target can be to drop guice even  if it doesnt happen before v6.

**Comment by gnodet** [comment]
I would not invest time to migrate to sisu.  I would rather migrate the remaining v3 services to the [DI framework API](https://github.com/apache/maven/tree/master/api/maven-api-di/src/main/java/org/apache/maven/api/di) and [implementation](https://github.com/apache/maven/tree/master/maven-di).

**Reply by gnodet** [reply]
FTR, the current DI framework has a visibility problem, see https://issues.apache.org/jira/browse/MNG-8139

**Reply by kwin** [reply]
I agree that for Maven4  Core/Plugins the DI framework API is the way to go. However there is one part missing: Configuration via XML (and type conversion). This is tracked in https://issues.apache.org/jira/browse/MNG-8343, because right now there is only Plexus API without any alternative.

**Comment by ascheman** [comment]
The solution to this discussion will be implemented with #44 (see the upcoming PR #76 also). 
But feel free to continue discussing possible solutions here.
